File: cmd/track.ts
// `f5kb track` — indexes a dump into the SQLite overview and reports
// new/changed/unchanged/removed. Behavior reference: track_articles.ts main().
// The DB schema/upsert come from lib/track/db.ts (byte-identical). Returns 0.
//
// Flags:
//   --dump=DIR     dump directory (default outputs/dump)
//   --db=FILE      SQLite file (default <dump>/../articles.db)
//   --types="A,B"  only index these type subdirs
//   --run-id=ID    label for this run (default current ISO timestamp)
//   --json         print the run summary as JSON to STDOUT

import { type ParsedArgs } from "../lib/args.ts";
import { flagBool, flagList, flagStr } from "../lib/args.ts";
import { type Logger, makeLogger } from "../lib/logger.ts";
import { trackDump } from "../lib/track/db.ts";
import { Changelog, changelogPathFromFlag } from "../lib/changelog.ts";

export async function run(args: ParsedArgs, logger: Logger): Promise<number> {
  const flags = args.flags;

  const dump = flagStr(flags, "dump", "outputs/dump")!;
  const db = flagStr(flags, "db");
  const types = flagList(flags, "types");
  const excludeTypes = flagList(flags, "exclude-types");
  // Pin one runId so the changelog records, the runs row, and the changes rows all
  // share it (trackDump would otherwise default to its own fresh timestamp).
  const runId = flagStr(flags, "run-id") ?? new Date().toISOString();
  const asJson = flagBool(flags, "json");

  // In --json mode, floor the logger at warn so the human info summary is dropped
  // (the JSON on STDOUT is the only payload) while skip-unreadable warnings still
  // surface on STDERR. Otherwise log the full human summary.
  const trackLogger = asJson ? makeLogger({ level: "warn", json: false, scope: "track" }) : logger;

  const changelogPath = changelogPathFromFlag(flags["changelog"], dump);
  const changelog = new Changelog(changelogPath, runId);

  const summary = await trackDump({
    dump,
    db,
    types,
    excludeTypes,
    runId,
    logger: trackLogger,
    changelog,
  });

  await changelog.flush();

  if (asJson) {
    console.log(JSON.stringify(summary, null, 2));
  } else if (changelogPath) {
    logger.info(`Changelog: ${changelogPath}`);
  }
  return 0;
}

File: f5kb/cmd/track.py
"""f5kb track subcommand."""

from __future__ import annotations

import json
import sys

import click

from f5kb.lib.changelog import Changelog, changelog_path_from_flag
from f5kb.track.db import track_dump


@click.command()
@click.option("--dump", "dump_dir", default="outputs/dump", show_default=True)
@click.option("--db", default=None)
@click.option("--types", default=None)
@click.option("--exclude-types", default=None)
@click.option("--run-id", default=None)
@click.option("--changelog", "changelog_flag", default=None, is_flag=False, flag_value="")
@click.option("--json", "as_json", is_flag=True)
@click.pass_context
def track_cmd(ctx, dump_dir, db, types, exclude_types, run_id, changelog_flag, as_json):
    """Index a dump into the SQLite overview; report new/changed/removed."""
    from f5kb.lib.logger import make_logger
    log = ctx.obj["logger"]
    if as_json:
        log = make_logger(level="warn", json_mode=False, scope="track")
    include_types = [t.strip() for t in types.split(",")] if types else None
    excl_types = [t.strip() for t in exclude_types.split(",")] if exclude_types else None
    import datetime
    run = run_id or datetime.datetime.now(datetime.timezone.utc).isoformat().replace("+00:00", "Z")
    cl_path = changelog_path_from_flag(changelog_flag, dump_dir)
    changelog = Changelog(cl_path, run)

    summary = track_dump(
        dump_dir, db_path=db, types=include_types, exclude_types=excl_types,
        run_id=run, logger=log, changelog=changelog,
    )
    changelog.flush()
    if as_json:
        import dataclasses
        click.echo(json.dumps(dataclasses.asdict(summary), indent=2))
    elif cl_path:
        log.info(f"Changelog: {cl_path}")
